dvc/remote/base.py

```diff
@@ -547,11 +547,51 @@ def download(
         if to_info.scheme != "local":
             raise NotImplementedError
 
-        logger.debug("Downloading '{}' to '{}'".format(from_info, to_info))
+        if self.isdir(from_info):
+            return self._download_dir(
+                from_info, to_info, name, no_progress_bar, file_mode, dir_mode
+            )
+        return self._download_file(
+            from_info, to_info, name, no_progress_bar, file_mode, dir_mode
+        )
 
-        name = name or to_info.name
+    def _download_dir(
+        self, from_info, to_info, name, no_progress_bar, file_mode, dir_mode
+    ):
+        file_to_infos = (
+            to_info / file_to_info.relative_to(from_info)
+            for file_to_info in self.walk_files(from_info)
+        )
 
+        with ThreadPoolExecutor(max_workers=self.JOBS) as executor:
```

Comment:
`download`/`upload` are called inside `download` with `ThreadPoolExecutor`. What we are having here is:
(eg)
`pull` -> `download` (JOBS threads) -> `_download_dir(JOBS threads)` -> so we actually can make up to JOBS ^ 2 processes in case of pulling directories.
I would just remove it from here.

Reply:
My bad, in case of import-url it will behave as single `download` call and in case of pull/push we prepare plans anyway, and they rely on singular files from the cache.